ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs

```diff
@@ -171,105 +216,321 @@ data InitLog blk =
-  -> StreamAPI m blk
-  -> m (InitLog blk, LedgerDB' blk, Word64)
+  => Tracer m (ReplayGoal blk -> TraceReplayEvent blk)   -- ^ A replay tracer
+  -> Tracer m (TraceEvent blk)                           -- ^ A general tracer
+  -> SomeHasFS m                                         -- ^ The filesystem with the old-style snapshots
+  -> (forall s. Decoder s (ExtLedgerState blk ValuesMK)) -- ^ A decoder for the old-style snapshots
+  -> SomeHasFS m                                         -- ^ The filesystem with the new-style snapshots
+  -> OnDiskLedgerStDb m (ExtLedgerState blk) blk         -- ^ The backend handle
+  -> (forall s. Decoder s (ExtLedgerState blk EmptyMK))  -- ^ A decoder for the new-style snapshots
+  -> (forall s. Decoder s (HeaderHash blk))              -- ^ A decoder for header hashes
+  -> LedgerDbCfg (ExtLedgerState blk)                    -- ^ The ledger database configuration
+  -> m (ExtLedgerState blk ValuesMK)                     -- ^ Genesis ledger state
+  -> StreamAPI m blk                                     -- ^ An streaming API for the immutable database
+  -> Bool                                                -- ^ Whether to run both databases TODO: This is not yet implemented
+  -> m ((InitLog blk, InitLog blk), LedgerDB' blk, Word64)
 initLedgerDB replayTracer
              tracer
              hasFS
-             onDiskLedgerDbSt
              decLedger
+             newHasFS
+             onDiskLedgerDbSt
+             decNewLedger
              decHash
              cfg
-             _getGenesisLedger
-             streamAPI = do
-    snapshots <- listSnapshots hasFS
-    -- Here we'd get the dbhandle
-    --
-    -- We can use this dbhandle to fetch the restore points.
-    tryNewestFirst id snapshots
+             getGenesisLedger
+             streamAPI
+             runDual = do
+    (initLog, ledgerDb, tipPoint)                    <- oldInitLedgerDB
+                                                          getGenesisLedger
+                                                          hasFS
+                                                          decLedger
+                                                          decHash
+    (initLog', ledgerDb', tipPoint', replayTracer') <- newInitLedgerDB
+                                                          replayTracer
+                                                          tracer
+                                                          getGenesisLedger
+                                                          newHasFS
+                                                          decNewLedger
+                                                          decHash
+                                                          onDiskLedgerDbSt
+
+    (lgrDB, w) <- case compare tipPoint tipPoint' of
+        LT -> bringUpOldLedgerDB runDual               cfg streamAPI                  ledgerDb ledgerDb' tipPoint tipPoint'
```

Comment:
I wonder if we could unify theses cases and `initStartingWith`: we simply ignore block applications in one of the DB's if the DB tip is ahead of the block being applied.

Reply:
One option we discussed today:

```haskell
ledgerDbPush :: forall m c l blk
              . (ApplyBlock l blk, TickedTableStuff l, Monad m, c)
             => LedgerDbCfg l
             -> Ap m l blk c -> LedgerDB l -> m (LedgerDB l)
ledgerDbPush cfg ap db =
  case runDual db of
    RunOnDiskOnly -> do
      currentNew' <- applyBlockNew (ledgerDbCfg cfg) ap (ledgerDbChangelog db)
      db { ledgerDbChangelog = extendDbChangelog (stateSeqNo currentNew')
                                                 (trackingTablesToDiffs currentNew')
                                                 ledgerDbChangelog
         }
    RunBoth        -> do
      currentOld' <- ...
      currentNew' <- applyBlockNew (ledgerDbCfg cfg) ap (ledgerDbChangelog db)
      let oldLedgerDbCurrent = either unCheckpoint unCheckpoint . AS.head . ledgerDbCheckpoints $ db
      assert (currentOld' == oldLedgerDbCurrent oldLedgerDbCurrent currentNew')
      db { ledgerDbChangelog = extendDbChangelog (stateSeqNo currentNew')
                                                 (trackingTablesToDiffs currentNew')
                                                 ledgerDbChangelog
         , ...
         }
   

```